### ISSUE TITLE
rxjs/no-async-subscribe fails with Error: Unknown node type JSXElement.

### ISSUE THREAD

**Issue by felixfbecker** [cartant/eslint-plugin-rxjs#46]
I narrowed this down by disabling all rules except this one. At happens when linting https://github.com/sourcegraph/sourcegraph/blob/master/shared/src/actions/ActionItem.test.tsx

```
Error: Unknown node type JSXElement.
Occurred while linting /Users/felix/git/sourcegraph/shared/src/actions/ActionItem.test.tsx:99
    at Controller.traverse (/Users/felix/git/sourcegraph/node_modules/estraverse/estraverse.js:521:31)
    at Object.traverse (/Users/felix/git/sourcegraph/node_modules/estraverse/estraverse.js:713:27)
    at matches (/Users/felix/git/sourcegraph/node_modules/esquery/esquery.js:90:34)
    at matches (/Users/felix/git/sourcegraph/node_modules/esquery/esquery.js:76:30)
    at Function.matches (/Users/felix/git/sourcegraph/node_modules/esquery/esquery.js:104:32)
    at NodeEventGenerator.applySelector (/Users/felix/git/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:253:21)
    at NodeEventGenerator.applySelectors (/Users/felix/git/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
    at NodeEventGenerator.enterNode (/Users/felix/git/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
    at CodePathAnalyzer.enterNode (/Users/felix/git/sourcegraph/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:634:23)
    at /Users/felix/git/sourcegraph/node_modules/eslint/lib/linter/linter.js:936:32
```

The rest of the ESLint config can be found here: https://github.com/sourcegraph/eslint-config/blob/af296ed7ee18b0e7f05990868caa96698f56e54d/.eslintrc.js

**Comment by cartant** [comment]
`JSXElement` doesn't seem to be in the visitor keys that are used to configure `traverse`: https://github.com/cartant/eslint-etc/blob/master/source/configure-traverse.ts

https://github.com/typescript-eslint/typescript-eslint/blob/188b68979d114ba0957a076fe9c27d154223512c/packages/typescript-estree/src/visitor-keys.ts#L36-L40

What fun.

**Comment by felixfbecker** [comment]
I think I also found the same problem with `OptionalMemberExpression`
```

Occurred while linting /buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/cmd/precise-code-intel/src/bundle-manager/backend/database.test.ts:114
--
  | at Controller.traverse (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/estraverse/estraverse.js:521:31)
  | at Object.traverse (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/estraverse/estraverse.js:713:27)
  | at matches (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/esquery/esquery.js:90:34)
  | at matches (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/esquery/esquery.js:76:30)
  | at Function.matches (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/esquery/esquery.js:104:32)
  | at NodeEventGenerator.applySelector (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:253:21)
  | at NodeEventGenerator.applySelectors (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:283:22)
  | at NodeEventGenerator.enterNode (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/eslint/lib/linter/node-event-generator.js:297:14)
  | at CodePathAnalyzer.enterNode (/buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/eslint/lib/linter/code-path-analysis/code-path-analyzer.js:634:23)
  | at /buildkite/builds/buildkite-agent-675457b655-d9l5z-1/sourcegraph/sourcegraph/node_modules/eslint/lib/linter/linter.js:936:32


```

**Comment by cartant** [comment]
TBH, I don't know whether I'm willing to put in the effort to support these ESLint/TypeScript plugins. The parser is now barfing on the snippets in https://github.com/cartant/eslint-plugin-rxjs/blob/master/tests/rules/no-unbound-methods.ts which, AFAICT, are perfectly valid. TypeScript itself seems to have no problem with them. The fact that plugins are at the mercy of whatever `@typescript-eslint/parser` happens to be installed is pretty frustrating.

FWIW, the `query` calls - that use `traverse` internally - need to be replaced. There are three left. What's there would work - if the visitor keys were correct - and was easier to port from the TSLint implementation of the rules.

Given that the parser now breaks on snippets that are valid, haven't changed, used to work and are unrelated to the `query` refactoring, I'm putting this to one side, for now.

**Comment by cartant** [comment]
FWIW, the parser was breaking because I changed the 'virtual' test file's extension to `.tsx` (so that I could add a test for this issue): https://github.com/cartant/eslint-plugin-rxjs/commit/389976e13999dda3fd3a9ac4f2f97ba94ef053d0

The test passes so ... 🤷‍♂ And the `no-async-subscribe` does not used the `query`, so the `traverse` failure is not coming from within the rule's implementation. Which means the problem is not with this package, AFAICT.

**Comment by felixfbecker** [comment]
Sorry to hear this is so much trouble. Maybe @bradzacher has an idea?

**Comment by cartant** [comment]
No worries. I'll have a another look at it when I'm less grumpy.

**Comment by bradzacher** [comment]
Note that whilst `JSXElement` isn't listed in the list of visitor keys in the parser, it is included via `eslint-visitor-keys`.

We use `eslint-visitor-keys`'s `unionWith` method to make sure we get the base nodes, and override what's required.

It'll be missing the `typeParameters` property though, but that's not causing a bug in this case (though I raised https://github.com/typescript-eslint/typescript-eslint/pull/1839) to add it in.

https://github.com/typescript-eslint/typescript-eslint/blob/188b68979d114ba0957a076fe9c27d154223512c/packages/typescript-estree/src/visitor-keys.ts#L1-L3

https://github.com/eslint/eslint-visitor-keys/blob/0516192692d32b22509d3b34705dc13ec713f996/lib/visitor-keys.json#L148-L152

---

> The fact that plugins are at the mercy of whatever @typescript-eslint/parser happens to be installed is pretty frustrating

Definitely. We try to avoid breaking changes without a major bump for this reason. But between
incomplete test coverage, and unknown cases, sometimes things slip through.
If there's something that broke between minors, then let us know so we can look into it.

In terms of our release schedule, we release a new version to the `latest` tag on npm every monday morning (PST), and we release a new version to the `canary` tag on every single commit to master.

---

As for why ESQuery fails - I don't know exactly. ESQuery doesn't actually care about what you put into the query - it's just an engine to parse and apply a query.

I can tell ESQuery to query for `MyUnknonwNode[someRandomTag="foo"]`, and ESQuery will parse this as is, convert it to a query (matches node with `type === "MyUnknonwNode"` and attribute `somRandomTag === "foo"`

You can play with it here: http://estools.github.io/esquery/

Could you dump the log of a `--debug` run with ESLint so I can see the full context?

**Comment by cartant** [comment]
@bradzacher Thanks for the information.

@felixfbecker I've created a separate repo that can be used to reproduce - using specific versions - ESLint plugin problems. And I've created a repro - which passes - for this issues: https://github.com/cartant/eslint-issues/commit/19e1ec0672a6efa5fa6f64deb6aa92e5329fdb55

Hopefully, this approach should make determining the cause of problems a little easier.

**Comment by cartant** [comment]
@felixfbecker It's possible that this could be related to the `eslint-etc` utils having `esquery` as a dependency. It should likely be a `peerDependency` - in which case the [configuration](https://github.com/cartant/eslint-etc/blob/98e1d38110d4e28f71f918aa433d1867d3c7daf4/source/query.ts#L15) of [`estraverse`](https://github.com/cartant/eslint-etc/blob/98e1d38110d4e28f71f918aa433d1867d3c7daf4/source/configure-traverse.ts#L14-L28) should be unnecessary, as the `typescript-eslint` infrastructure should have already done it. (And, if the both dependencies resolve to the same version, it's possible that the configuration in `eslint-etc` is stomping on `typescript-eslint`'s configuration.)

Why this would happen if `no-async-subscribe` is the only rule you are using, IDK, as it doesn't call the `query` util. You could try commenting this out, to see what happens: https://github.com/cartant/eslint-etc/blob/master/source/query.ts#L14-L17

**Comment by felixfbecker** [comment]
These are good hints!

Here's what `esquery` looks like in my node_modules:

```
> npm ls esquery              
/Users/felix/git/sourcegraph
+-- @sourcegraph/eslint-config@0.12.0
| `-- eslint-plugin-etc@0.0.1-beta.25
|   +-- @phenomnomnominal/tsquery@4.0.0
|   | `-- esquery@1.0.1  deduped
|   `-- eslint-etc@0.0.8
|     `-- esquery@1.0.1  deduped
+-- eslint@6.8.0
| `-- esquery@1.0.1 
`-- web-ext@4.1.0
  `-- addons-linter@1.20.0
    +-- eslint@5.16.0
    | `-- esquery@1.0.1  deduped
    `-- eslint-plugin-no-unsafe-innerhtml@1.0.16
      `-- eslint@3.19.0
        `-- esquery@1.0.1  deduped
```

I noticed that is an outdated version, so I wanted to see what happens if I update it:

```
> npm ls esquery              
/Users/felix/git/sourcegraph
+-- @sourcegraph/eslint-config@0.12.0
| `-- eslint-plugin-etc@0.0.1-beta.25
|   +-- @phenomnomnominal/tsquery@4.0.0
|   | `-- esquery@1.0.1  deduped
|   `-- eslint-etc@0.0.8
|     `-- esquery@1.0.1  deduped
+-- eslint@6.8.0
| `-- esquery@1.0.1 
`-- web-ext@4.1.0
  `-- addons-linter@1.20.0
    +-- eslint@5.16.0
    | `-- esquery@1.0.1  deduped
    `-- eslint-plugin-no-unsafe-innerhtml@1.0.16
      `-- eslint@3.19.0
        `-- esquery@1.0.1  deduped
```

Now if an update _is_ required, then of course some package should have bumped their requirement (which package though I don't know).

Unfortunately, that update makes ESLint run forever and never return (been running for >20min).

**Comment by cartant** [comment]
@felixfbecker FWIW, I was getting a running-forever test until I made the change in this commit: https://github.com/cartant/eslint-plugin-rxjs/commit/ee48d85345993a04108b89e2b434eb335d90769b

I made that change, yesterday, so it's there is no released version of the package with that change. You might want to try building the plugin using the repo's current source.

I can have another look at this tomorrow.

**Comment by cartant** [comment]
@felixfbecker This might be fixed in ` 0.0.2-beta.17`.

The problem was that the `no-async-subscribe` rule uses `has` in its selector and that effects a traversal within `esquery`. So unless `estraverse` happened to have been configured through the execution of another rule in `eslint-plugin-rxjs`, `estraverse` would not have had its `VisitorKeys` configured for the TypeScript nodes that it would have encountered. (Note that this was not only dependent upon rule use and rule ordering, but also depended upon the AST node that were encountered!)

The (re)configuration of `estraverse` is no longer done inside the `query` call. It's now done at the module level for each rule that requires traversal. (Module-level configuration is the worst, but there's not much that I can do about it, as `estraverse` has a module-level configuration API.)

FWIW, `typesccript-eslint` does not have to configure `estraverse` because ESLint performs its own traversal and calls the `parse` and `matches` functions in `esquery` - and the latter traverses only if `has` is used.

**Comment by Delagen** [comment]
Latest beta 18 version no-async-subscribe hangs in CI.. Disabled it for a while

**Comment by cartant** [comment]
@Delagen Without a failing reproduction in this repo, I won't be spending further time with this: https://github.com/cartant/eslint-issues

If someone is able to create a reproduction using specific versions, I will have a look at it.

**Comment by cartant** [comment]
This should be fixed in `0.0.2-beta.19`. That release removes `:has` from this rule's selector and from the selectors of a bunch of other rules.
